File: mkdoc.pl
#!/usr/bin/perl

# The RiveScript.JS docs are somewhat JavaDoc-ish, but not quite.
# This is just a quick Perl script to make the docs user friendly.

my $inJavadoc = 0;
my @blocks = { type => "text", text => [] };

open (my $fh, "<", "bin/RiveScript.js");
while (my $line = <$fh>) {
	chomp $line;
	$line =~ s/^[\s\t]+//g;

	if ($line =~ /^\/\*\*/) {
		$inJavadoc = 1;
		next;
	} elsif ($line =~ /\*\//) {
		$inJavadoc = 0;
		push @blocks, {
			type => "text",
			text => [],
		};
		next;
	}

	if ($inJavadoc) {
		if ($line !~ /^\*/) {
			warn "Weird error: line doesn't begin with '* '";
			next;
		}
		$line =~ s/^\*\s*//;

		# Class lines.
		if ($line =~ /^(JsRiveObjects|RiveScript)/ && $blocks[-1]->{type} eq "text") {
			print "Found package: $1\n";
			$blocks[-1]->{type} = "package";
			$blocks[-1]->{package} = $1;
			$blocks[-1]->{declare} = $line;
		}

		# Function lines
		elsif ($line =~ /^(void|string|float|private|int|bool|data)/ && $blocks[-1]->{type} eq "text") {
			$blocks[-1]->{type}    = "function";
			$blocks[-1]->{declare} = $line;
		}

		# Text
		else {
			push @{$blocks[-1]->{text}}, $line;
		}
	}
}
close($fh);
undef $fh;

print Dumper(\@blocks); use Data::Dumper;

# Write to the docs folder.
foreach my $block (@blocks) {
	# A package block?
	if ($block->{type} eq "package") {
		# Close out the old file?
		if (defined $fh) {
			print {$fh} html_footer();
		}

		# Write a new file.
		my $package = $block->{package};
		open ($fh, ">", "./docs/$package.html");
		print {$fh} html_header($block) . "<hr>\n\n";
	}
	elsif ($block->{type} eq "function") {
		next if scalar @{$block->{text}} == 0;
		print {$fh} "<h2>$block->{declare}</h2>\n\n"
			. join("<br>\n", @{$block->{text}}) . "<hr>\n\n";
	}
	else {
		next if scalar @{$block->{text}} == 0;
		print {$fh} join("<br>\n", @{$block->{text}}) . "<hr>\n\n";
	}
}

print {$fh} html_footer();

sub html_header {
	my $block = shift;
	return <<"EOF"
<!DOCTYPE html>
<html>
<head>
<title>$block->{package}</title>
<link rel="stylesheet" type="text/css" href="docs.css">
</head>
<body>

<h1>$block->{package}</h1>

<hr>

<h1>Methods</h1>

<h2>$block->{declare}</h2>
EOF
	. join("<br>\n", @{$block->{text}}) . "\n\n";
}

sub html_footer {
	return "</body>\n</html>\n";
}

File: mkdoc.py
#!/usr/bin/env python3

"""Generate documentation from the RiveScript source files."""

import os
import os.path
import markdown

def main():
    # Scan for JavaScript modules.
    modules = []
    for path, dirs, files in os.walk("src"):
        for f in files:
            module = os.path.join(path, f) \
                .replace("src/", "") \
                .replace(".js", "") \
                .replace("/", ".")
            modules.append(dict(
                name=module,
                path=os.path.join(path, f),
            ))

    # Read their documentation.
    for mod in modules:
        md = []  # Markdown output, in sections
        in_doc = False
        buf = []
        fh = open(mod["path"], "r")
        indent = 0
        for line in fh:
            stripped = line.strip()
            if stripped.startswith("/**") and not in_doc:
                in_doc = True
                indent = line.index("/**")
                continue
            elif stripped.startswith("*/") and in_doc:
                in_doc = False
                md.append(to_markdown(buf))
                buf = []
                continue

            if in_doc:
                line = line[indent:].rstrip()
                buf.append(line)
        fh.close()

        # Write the Markdown file.
        print("Write docs for module: {}".format(mod["name"]))
        fh = open("docs/{}.md".format(mod["name"]), "w")
        fh.write("\n\n".join(md))
        fh.close()

        # Render the HTML from the Markdown.
        render_markdown(
            mod["name"],
            "docs/html/{}.html".format(mod["name"]),
            "\n\n".join(md)
        )


def to_markdown(buf):
    """Convert documentation data to markdown."""
    output = []

    # Consolodate the header.
    header = []
    while len(buf) > 0 and buf[0].strip():
        header.append(buf.pop(0).strip())

    # Headers that aren't functions are always shown as H1.
    function_types = ["string", "void", "data", "private", "object", "int",
        "bool", "Promise", "[]string", "async"]
    is_function = False
    for t in function_types:
        if header[0].startswith(t):
            is_function = True
            break

    if not is_function:
        output.append("# {}".format(" ".join(header)))
    else:
        output.append("## {}".format(" ".join(header)))

    for line in buf:
        output.append(line)

    return "\n".join(output).strip()


def render_markdown(title, path, text):
    """Render Markdown to HTML."""
    fh = open(path, "w")

    # HTML header.
    fh.write("""<!DOCTYPE html>
<html>
<head>
<title>{title}</title>
<link rel="stylesheet" type="text/css" href="md.css">
</head>
<body>

{text}

</body>
</html>""".format(
        title=title,
        text=markdown.markdown(text,
            extensions=[
                "fenced_code",
                "smart_strong",
                "codehilite",
                "sane_lists",
            ],
            safe_mode="escape",
        ),
    ))

    fh.close()


if __name__ == "__main__":
    main()
